guides/introduction/Embedded Schemas.md

```diff
@@ -92,57 +87,124 @@ end
 
 ```elixir
 defmodule UserProfile do
   # ...
 
-  def changeset(profile, attrs \\ %{}) do
+  def changeset(%UserProfile{} = profile, attrs \\ %{}) do
     profile
-    |> cast(attrs, [:age, :favorite_color, :avatar_url])
-    |> validate_required([:age, :favorite_color])
+    |> cast(attrs, [:online, :dark_mode, :visibility])
+    |> validate_required([:online, :visibility])
   end
 end
+
+profile = %UserProfile{}
+UserProfile.changeset(profile, %{online: true, visibility: :public})
 ```
 
-On the user side, you also define a `changeset/2` function, and then you use `cast_embed/3` to invoke the `UserProfile` changeset:
+Meanwhile, the User changeset function can require it's own validations without worrying about the details of the UserProfile changes because it can pass that responsibility to UserProfile via `cast_embed/3`. A validation failure in an embed will cause the parent changeset to be invalid, even if the parent changeset itself had no errors.
 
 ```elixir
 defmodule User do
-  use Ecto.Schema
-
   # ...
 
   def changeset(user, attrs \\ %{}) do
+    user
+    |> cast(attrs, [:full_name, :email, :avatar_url])
+    |> cast_embed(:profile, required: true)
+  end
+end
+
+changeset = User.changeset(%User{}, %{profile: %{online: true}})
+changeset.valid? # => false; "visibility can't be blank"
+changeset = User.changeset(%User{}, %{profile: %{online: true, visibility: :public}})
+changeset.valid? # => true
+```
+
+In situations where you have kept the embedded schema within the parent module, e.g., you have not extracted a UserProfile, you can still have custom changeset functions for the embedded data within the parent schema.
+
+```elixir
+defmodule User do
+  use Ecto.Schema
+
+  schema "users" do
+    field :full_name, :string
+    field :email, :string
+    field :avatar_url, :string
+    field :confirmed_at, :naive_datetime
+
+    embeds_one :profile, Profile do
+      field :online, :boolean
+      field :dark_mode, :boolean
+      field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
+    end
+
+    timestamps()
+  end
+
+  def changeset(%User{} = user, attrs \\ %{}) do
     user
     |> cast(attrs, [:full_name, :email])
-    # By default it calls UserProfile.changeset/2, pass the :with option to change it
-    |> cast_embed(:user_profile, required: true)
+    |> cast_embed(:profile, required: true, with: &profile_changeset/2)
+  end
+
+  def profile_changeset(profile, attrs \\ %{}) do
+    profile
+    |> cast(attrs, [:online, :dark_mode, :visibility])
+    |> validate_required([:online, :visibility])
   end
 end
+
+changeset = User.changeset(%User{}, %{profile: %{online: true, visibility: :public}})
+changeset.valid? # => true
 ```
+
+### Querying embedded data
```

Comment:
I went ahead and wrote this query section even though I haven't gotten the `where` queries to work yet (which I mention above in the unresolved comment thread we are chatting in). The `select` queries work just fine. I think once that final issue above gets resolved it's ready to go! (I might do more edit passes, though, because I never like what I write, lol, XD). Thank you for taking the time to help me as I wrote this. It's been an honor <3. I'd love to write more if there's other guides and stuff you've been wanting to write but haven't had time for, even outside ecto, etc. :D